server/migrations/2020-06-30-135809_remove_mat_views/up.sql

```diff
@@ -0,0 +1,576 @@
+  END IF;
+
+  return null;
+end $$;
+
+-- Post
+
+create table post_aggregates_fast as select * from post_aggregates_view;
+alter table post_aggregates_fast add column fast_id serial primary key;
+
+create index idx_post_aggregates_fast_id on post_aggregates_fast (id);
+
+-- For the hot rank resorting
+create index idx_post_aggregates_fast_hot_rank_published on post_aggregates_fast (hot_rank desc, published desc);
+
+create view post_fast_view as 
+select
+ap.*,
+u.id as user_id,
+coalesce(pl.score, 0) as my_vote,
+(select cf.id::bool from community_follower cf where u.id = cf.user_id and cf.community_id = ap.community_id) as subscribed,
+(select pr.id::bool from post_read pr where u.id = pr.user_id and pr.post_id = ap.id) as read,
+(select ps.id::bool from post_saved ps where u.id = ps.user_id and ps.post_id = ap.id) as saved
+from user_ u
+cross join (
+  select
+  pa.*
+  from post_aggregates_fast pa
+) ap
+left join post_like pl on u.id = pl.user_id and ap.id = pl.post_id
+
+union all
+
+select 
+paf.*,
+null as user_id,
+null as my_vote,
+null as subscribed,
+null as read,
+null as saved
+from post_aggregates_fast paf
+;
+
+drop trigger refresh_post on post;
+
+create trigger refresh_post
+after insert or update or delete
+on post
+for each row
+execute procedure refresh_post();
+
+-- Sample select
+-- select id, name from post_fast_view where name like 'test_post' and user_id is null;
+-- Sample insert 
+-- insert into post(name, creator_id, community_id) values ('test_post', 2, 2);
+-- Sample delete
+-- delete from post where name like 'test_post';
+-- Sample update
+-- update post set community_id = 4  where name like 'test_post';
+create or replace function refresh_post()
+returns trigger language plpgsql
+as $$
+begin
+  IF (TG_OP = 'DELETE') THEN
+    delete from post_aggregates_fast where id = OLD.id;
+
+    -- Update community number of posts
+    update community_aggregates_fast set number_of_posts = number_of_posts - 1 where id = OLD.community_id;
+  ELSIF (TG_OP = 'UPDATE') THEN
+    delete from post_aggregates_fast where id = OLD.id;
+    insert into post_aggregates_fast select * from post_aggregates_view where id = NEW.id;
+  ELSIF (TG_OP = 'INSERT') THEN
+    insert into post_aggregates_fast select * from post_aggregates_view where id = NEW.id;
+
+    -- Update that users number of posts, post score
+    delete from user_fast where id = NEW.creator_id;
+    insert into user_fast select * from user_view where id = NEW.creator_id;
+  
+    -- Update community number of posts
+    update community_aggregates_fast set number_of_posts = number_of_posts + 1 where id = NEW.community_id;
+
+    -- Update the hot rank on the post table
+    -- TODO this might not correctly update it, using a 1 week interval
+    update post_aggregates_fast as paf
+    set hot_rank = pav.hot_rank 
+    from post_aggregates_view as pav
+    where paf.id = pav.id  and (pav.published > ('now'::timestamp - '1 week'::interval));
```

Comment:
hotRank is a time based decay function, but its static data in this table / cache. So not only do we need to add a hot rank for the new post that got added, we need to compute new hot ranks for all the others it could possibly be compared to. (I used a week interval to do this decay, otherwise it would update every single previous post, possibly years in the past. 

The only danger here is if your instance recieves 0 posts / week.

Reply:
Wouldn't we use `paf.id = pav.id or (hot rank)` in that case?  With the and we are just comparing the new post to itself.